decidim-proposals/spec/controllers/decidim/proposal_endorsements_controller_spec.rb

```diff
@@ -0,0 +1,301 @@
+
+require "spec_helper"
+
+module Decidim
+  module Proposals
+    describe ProposalEndorsementsController, type: :controller do
+      routes { Decidim::Proposals::Engine.routes }
+
+      let(:proposal) { create(:proposal, feature: feature) }
+      let(:user) { create(:user, :confirmed, organization: feature.organization) }
+
+      let(:params) do
+        {
+          proposal_id: proposal.id,
+          feature_id: feature.id,
+          participatory_process_slug: feature.participatory_space.slug
+        }
+      end
+
+      before do
+        request.env["decidim.current_organization"] = feature.organization
+        request.env["decidim.current_feature"] = feature
+        sign_in user
+      end
+
+      describe "As User" do
+        context "when endorsements are enabled" do
+          let(:feature) do
+            create(:proposal_feature, :with_endorsements_enabled)
+          end
+
+          it "allows endorsements" do
+            expect do
+              post :create, format: :js, params: params
+            end.to change { ProposalEndorsement.count }.by(1)
+
+            expect(ProposalEndorsement.last.author).to eq(user)
+            expect(ProposalEndorsement.last.proposal).to eq(proposal)
+          end
+
+          context "when requesting user identities without belonging to any user_group" do
+            it "onlies return the user identity endorse button" do
+              get :identities, params: params
+
+              expect(response).to have_http_status(:ok)
+              expect(assigns[:to_endorse_groups]).to be_empty
+              expect(assigns[:to_unendorse_groups]).to be_empty
+            end
+          end
+        end
+
+        context "when endorsements are disabled" do
+          let(:feature) do
+            create(:proposal_feature, :with_endorsements_disabled)
+          end
+
+          it "does not allow endorsing" do
+            expect do
+              post :create, format: :js, params: params
+            end.not_to change { ProposalEndorsement.count }
+
+            expect(flash[:alert]).not_to be_empty
+            expect(response).to have_http_status(302)
+          end
+
+          context "when requesting user identities" do
+            it "raises exception" do
+              get :identities, params: params
+              expect(response).to have_http_status(302)
+            end
+          end
+        end
+
+        context "when endorsements are enabled but endorsements are blocked" do
+          let(:feature) do
+            create(:proposal_feature, :with_endorsements_enabled, :with_endorsements_blocked)
+          end
+
+          it "does not allow endorsing" do
+            expect do
+              post :create, format: :js, params: params
+            end.not_to change { ProposalEndorsement.count }
+
+            expect(flash[:alert]).not_to be_empty
+            expect(response).to have_http_status(302)
+          end
+          context "when requesting user identities" do
```

Comment:
Missing extra line between `end` and `context`. Same for all this file. There should be an extra line between all `end`s and next `context`, `describe` or `it`, to make the blocks clearer for the developer